{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "DNA Sequencing Classifier using K mer Counting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#importing Common Libraries\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 4380 entries, 0 to 4379\n",
      "Data columns (total 2 columns):\n",
      "sequence    4380 non-null object\n",
      "class       4380 non-null int64\n",
      "dtypes: int64(1), object(1)\n",
      "memory usage: 68.6+ KB\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "sequence    object\n",
       "class        int64\n",
       "dtype: object"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Basic analysis of given dataset\n",
    "data = pd.read_table(\"human_data.txt\")\n",
    "data.info()\n",
    "data.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence</th>\n",
       "      <th>class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>ATGCCCCAACTAAATACTACCGTATGGCCCACCATAATTACCCCCA...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ATGAACGAAAATCTGTTCGCTTCATTCATTGCCCCCACAATCCTAG...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ATGTGTGGCATTTGGGCGCTGTTTGGCAGTGATGATTGCCTTTCTG...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>ATGTGTGGCATTTGGGCGCTGTTTGGCAGTGATGATTGCCTTTCTG...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>ATGCAACAGCATTTTGAATTTGAATACCAGACCAAAGTGGATGGTG...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            sequence  class\n",
       "0  ATGCCCCAACTAAATACTACCGTATGGCCCACCATAATTACCCCCA...      4\n",
       "1  ATGAACGAAAATCTGTTCGCTTCATTCATTGCCCCCACAATCCTAG...      4\n",
       "2  ATGTGTGGCATTTGGGCGCTGTTTGGCAGTGATGATTGCCTTTCTG...      3\n",
       "3  ATGTGTGGCATTTGGGCGCTGTTTGGCAGTGATGATTGCCTTTCTG...      3\n",
       "4  ATGCAACAGCATTTTGAATTTGAATACCAGACCAAAGTGGATGGTG...      3"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.head(5)"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "Converting the given sequences of DNA in to vectors of fixed length. for default, we are considering 6 characters, as many organisms has this pattern"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6\n"
     ]
    }
   ],
   "source": [
    "#feature engineering to convert the DNA sequence to a fixed ength vecotr called K-mer\n",
    "size = int(input())\n",
    "def Kmers(sequence,size=6):\n",
    "    return[sequence[x:x+size].lower() for x in range(len(sequence)-size+1)]\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "data['words'] = data.apply(lambda x: Kmers(x['sequence']),axis=1)\n",
    "data=data.drop('sequence',axis=1)\n",
    "#Dropping the intital sequence after modification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>class</th>\n",
       "      <th>words</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>[atgccc, tgcccc, gcccca, ccccaa, cccaac, ccaac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>[atgaac, tgaacg, gaacga, aacgaa, acgaaa, cgaaa...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>[atgtgt, tgtgtg, gtgtgg, tgtggc, gtggca, tggca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>[atgtgt, tgtgtg, gtgtgg, tgtggc, gtggca, tggca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>[atgcaa, tgcaac, gcaaca, caacag, aacagc, acagc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>[atgtgt, tgtgtg, gtgtgg, tgtggc, gtggca, tggca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>[atgaag, tgaaga, gaagat, aagatt, agattg, gattg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>[atgcaa, tgcaac, gcaaca, caacag, aacagc, acagc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>8</td>\n",
       "      <td>3</td>\n",
       "      <td>[atgaag, tgaaga, gaagat, aagatt, agattg, gattg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>[atgtgt, tgtgtg, gtgtgg, tgtggc, gtggca, tggca...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   class                                              words\n",
       "0      4  [atgccc, tgcccc, gcccca, ccccaa, cccaac, ccaac...\n",
       "1      4  [atgaac, tgaacg, gaacga, aacgaa, acgaaa, cgaaa...\n",
       "2      3  [atgtgt, tgtgtg, gtgtgg, tgtggc, gtggca, tggca...\n",
       "3      3  [atgtgt, tgtgtg, gtgtgg, tgtggc, gtggca, tggca...\n",
       "4      3  [atgcaa, tgcaac, gcaaca, caacag, aacagc, acagc...\n",
       "5      3  [atgtgt, tgtgtg, gtgtgg, tgtggc, gtggca, tggca...\n",
       "6      3  [atgaag, tgaaga, gaagat, aagatt, agattg, gattg...\n",
       "7      3  [atgcaa, tgcaac, gcaaca, caacag, aacagc, acagc...\n",
       "8      3  [atgaag, tgaaga, gaagat, aagatt, agattg, gattg...\n",
       "9      3  [atgtgt, tgtgtg, gtgtgg, tgtggc, gtggca, tggca..."
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#separating the K mers so that, we get distinct gene codlets, which makes easy to identitfy the errors and critical points in the Geneome\n",
    "text_data=list(data['words'])\n",
    "for i in range(len(text_data)):\n",
    "    text_data[i]=' '.join(text_data[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "atgccc tgcccc gcccca ccccaa cccaac ccaact caacta aactaa actaaa ctaaat taaata aaatac aatact atacta tactac actacc ctaccg taccgt accgta ccgtat cgtatg gtatgg tatggc atggcc tggccc ggccca gcccac cccacc ccacca caccat accata ccataa cataat ataatt taatta aattac attacc ttaccc tacccc accccc ccccca ccccat cccata ccatac catact atactc tactcc actcct ctcctt tcctta ccttac cttaca ttacac tacact acacta cactat actatt ctattc tattcc attcct ttcctc tcctca cctcat ctcatc tcatca catcac atcacc tcaccc caccca acccaa cccaac ccaact caacta aactaa actaaa ctaaaa taaaaa aaaaat aaaata aaatat aatatt atatta tattaa attaaa ttaaac taaaca aaacac aacaca acacaa cacaaa acaaac caaact aaacta aactac actacc ctacca taccac accacc ccacct caccta acctac cctacc ctacct tacctc acctcc cctccc ctccct tccctc ccctca cctcac ctcacc tcacca caccaa accaaa ccaaag caaagc aaagcc aagccc agccca gcccat cccata ccataa cataaa ataaaa taaaaa aaaaat aaaata aaataa aataaa ataaaa taaaaa aaaaaa aaaaat aaaatt aaatta aattat attata ttataa tataac ataaca taacaa aacaaa acaaac caaacc aaaccc aaccct accctg ccctga cctgag ctgaga tgagaa gagaac agaacc gaacca aaccaa accaaa ccaaaa caaaat aaaatg aaatga aatgaa atgaac tgaacg gaacga aacgaa acgaaa cgaaaa gaaaat aaaatc aaatct aatctg atctgt tctgtt ctgttc tgttcg gttcgc ttcgct tcgctt cgcttc gcttca cttcat ttcatt tcattc cattca attcat ttcatt tcattg cattgc attgcc ttgccc tgcccc gccccc ccccca ccccac cccaca ccacaa cacaat acaatc caatcc aatcct atccta tcctag\n"
     ]
    }
   ],
   "source": [
    "print(text_data[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([4, 4, 3, ..., 6, 6, 6], dtype=int64)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Allocating data to the testing and training of model\n",
    "y_data = data.iloc[:,0].values\n",
    "y_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Considering the Genomic sequence as a Language, we interrupt the codlets as words so that, using NLP we bag them into clusters for processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4380, 232414)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "from sklearn.feature_extraction.text import CountVectorizer\n",
    "CV=CountVectorizer(ngram_range=(4,4))\n",
    "x=CV.fit_transform(text_data)\n",
    "print(x.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x2002db33648>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX0AAAD3CAYAAADxJYRbAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAARpklEQVR4nO3df4xl5X3f8ffHbKHBaQ02Y0J2lyypt05Imth0BCSOIjeb4sVYXlQZCVqFlYu7qoobJ04Tr+NKSIlcESUKsSUHaWOwoXIhLk3EtqYhK2zXShsIA8b88Nphigk7WX5MBMZJSGJv/M0f91n5evbuzM7cmXuXPO+XNLrnfJ/nnvMdNHzus+eeO5OqQpLUh1dMuwFJ0uQY+pLUEUNfkjpi6EtSRwx9SeqIoS9JHVkx9JPcnOS5JI+OGPtPSSrJWW0/ST6cZD7Jw0kuGJq7O8nj7Wv3+n4bkqQTcSIr/Y8DO5cWk2wF/iXw1FD5UmB7+9oD3Njmvhq4DrgIuBC4LsmZ4zQuSVq9TStNqKrPJdk2YugG4BeAO4dqu4Bba/CJr3uTnJHkHODNwIGqeh4gyQEGLyS3LXfus846q7ZtG3VqSdLxPPDAA39WVTOjxlYM/VGSvB3406r6QpLhoc3AoaH9hVY7Xn1Z27ZtY25ubi0tSlK3kvzJ8cZWHfpJTgc+AFwyanhErZapjzr+HgaXhjj33HNX254kaRlruXvnnwDnAV9I8iSwBXgwyXcxWMFvHZq7BTi8TP0YVbWvqmaranZmZuS/TiRJa7Tq0K+qR6rqtVW1raq2MQj0C6rqGWA/cHW7i+di4MWqehq4G7gkyZntDdxLWk2SNEEncsvmbcAfAq9PspDkmmWm3wU8AcwDvwX8B4D2Bu4vA/e3r186+qauJGlycjL/auXZ2dnyjVxJWp0kD1TV7KgxP5ErSR0x9CWpI4a+JHVkTR/OkiR9y7a9n9rQ4z95/WXrdixX+pLUEUNfkjpi6EtSRwx9SeqIoS9JHTH0Jakjhr4kdcTQl6SOGPqS1BFDX5I6YuhLUkcMfUnqiKEvSR0x9CWpI4a+JHXE0Jekjhj6ktQRQ1+SOrJi6Ce5OclzSR4dqv1qki8leTjJ7yY5Y2js/Unmk3w5yVuG6jtbbT7J3vX/ViRJKzmRlf7HgZ1LageAH6yqHwL+GHg/QJLzgSuBH2jP+c0kpyQ5BfgIcClwPnBVmytJmqAVQ7+qPgc8v6T2+1V1pO3eC2xp27uA26vqb6rqK8A8cGH7mq+qJ6rq68Dtba4kaYLW45r+vwX+d9veDBwaGltotePVj5FkT5K5JHOLi4vr0J4k6aixQj/JB4AjwCeOlkZMq2Xqxxar9lXVbFXNzszMjNOeJGmJTWt9YpLdwNuAHVV1NMAXgK1D07YAh9v28eqSpAlZ00o/yU7gfcDbq+qloaH9wJVJTktyHrAd+CPgfmB7kvOSnMrgzd7947UuSVqtFVf6SW4D3gyclWQBuI7B3TqnAQeSANxbVf++qh5L8kngiwwu+1xbVX/bjvNu4G7gFODmqnpsA74fSdIyVgz9qrpqRPmmZeZ/EPjgiPpdwF2r6k6StK78RK4kdcTQl6SOGPqS1BFDX5I6YuhLUkcMfUnqiKEvSR0x9CWpI4a+JHXE0Jekjhj6ktQRQ1+SOmLoS1JHDH1J6oihL0kdMfQlqSOGviR1xNCXpI4Y+pLUEUNfkjpi6EtSRwx9SerIiqGf5OYkzyV5dKj26iQHkjzeHs9s9ST5cJL5JA8nuWDoObvb/MeT7N6Yb0eStJwTWel/HNi5pLYXuKeqtgP3tH2AS4Ht7WsPcCMMXiSA64CLgAuB646+UEiSJmfF0K+qzwHPLynvAm5p27cAlw/Vb62Be4EzkpwDvAU4UFXPV9ULwAGOfSGRJG2wtV7TP7uqngZoj69t9c3AoaF5C612vPoxkuxJMpdkbnFxcY3tSZJGWe83cjOiVsvUjy1W7auq2aqanZmZWdfmJKl3aw39Z9tlG9rjc62+AGwdmrcFOLxMXZI0QWsN/f3A0TtwdgN3DtWvbnfxXAy82C7/3A1ckuTM9gbuJa0mSZqgTStNSHIb8GbgrCQLDO7CuR74ZJJrgKeAK9r0u4C3AvPAS8A7Aarq+SS/DNzf5v1SVS19c1iStMFWDP2quuo4QztGzC3g2uMc52bg5lV1J0laV34iV5I6YuhLUkcMfUnqiKEvSR0x9CWpI4a+JHXE0Jekjhj6ktQRQ1+SOmLoS1JHDH1J6oihL0kdMfQlqSOGviR1xNCXpI4Y+pLUEUNfkjpi6EtSRwx9SeqIoS9JHTH0Jakjhr4kdWSs0E/ys0keS/JoktuS/MMk5yW5L8njSX47yalt7mltf76Nb1uPb0CSdOLWHPpJNgM/DcxW1Q8CpwBXAr8C3FBV24EXgGvaU64BXqiq1wE3tHmSpAka9/LOJuA7kmwCTgeeBn4CuKON3wJc3rZ3tX3a+I4kGfP8kqRVWHPoV9WfAr8GPMUg7F8EHgC+WlVH2rQFYHPb3gwcas890ua/Zulxk+xJMpdkbnFxca3tSZJGGOfyzpkMVu/nAd8NvBK4dMTUOvqUZca+VajaV1WzVTU7MzOz1vYkSSOMc3nnJ4GvVNViVX0D+B3gR4Ez2uUegC3A4ba9AGwFaOOvAp4f4/ySpFUaJ/SfAi5Ocnq7Nr8D+CLwGeAdbc5u4M62vb/t08Y/XVXHrPQlSRtnnGv69zF4Q/ZB4JF2rH3A+4D3JplncM3+pvaUm4DXtPp7gb1j9C1JWoNNK085vqq6DrhuSfkJ4MIRc/8auGKc80mSxuMnciWpI4a+JHXE0Jekjhj6ktQRQ1+SOmLoS1JHDH1J6oihL0kdGevDWZJOHtv2fmpDj//k9Zdt6PE1Ga70Jakjhr4kdcTQl6SOGPqS1BFDX5I64t07UuPdL+qBK31J6oihL0kdMfQlqSOGviR1xNCXpI4Y+pLUEUNfkjoyVugnOSPJHUm+lORgkh9J8uokB5I83h7PbHOT5MNJ5pM8nOSC9fkWJEknatyV/oeA36uq7wN+GDgI7AXuqartwD1tH+BSYHv72gPcOOa5JUmrtObQT/KPgR8HbgKoqq9X1VeBXcAtbdotwOVtexdwaw3cC5yR5Jw1dy5JWrVxVvrfCywCH0vy+SQfTfJK4OyqehqgPb62zd8MHBp6/kKrfZske5LMJZlbXFwcoz1J0lLjhP4m4ALgxqp6I/CXfOtSzigZUatjClX7qmq2qmZnZmbGaE+StNQ4ob8ALFTVfW3/DgYvAs8evWzTHp8bmr916PlbgMNjnF+StEprDv2qegY4lOT1rbQD+CKwH9jdaruBO9v2fuDqdhfPxcCLRy8DSZImY9xfrfwfgU8kORV4AngngxeSTya5BngKuKLNvQt4KzAPvNTmSpImaKzQr6qHgNkRQztGzC3g2nHOJ0kaj5/IlaSOGPqS1BFDX5I6YuhLUkf8w+gnEf8wt6SN9vcq9A1NSVqel3ckqSOGviR1xNCXpI4Y+pLUEUNfkjpi6EtSRwx9SeqIoS9JHTH0Jakjhr4kdcTQl6SOGPqS1BFDX5I6YuhLUkcMfUnqyNihn+SUJJ9P8r/a/nlJ7kvyeJLfTnJqq5/W9ufb+LZxzy1JWp31WOm/Bzg4tP8rwA1VtR14Abim1a8BXqiq1wE3tHmSpAkaK/STbAEuAz7a9gP8BHBHm3ILcHnb3tX2aeM72nxJ0oSMu9L/DeAXgG+2/dcAX62qI21/AdjctjcDhwDa+IttviRpQtYc+kneBjxXVQ8Ml0dMrRMYGz7uniRzSeYWFxfX2p4kaYRxVvpvAt6e5EngdgaXdX4DOCPJ0T+4vgU43LYXgK0AbfxVwPNLD1pV+6pqtqpmZ2ZmxmhPkrTUmkO/qt5fVVuqahtwJfDpqvo3wGeAd7Rpu4E72/b+tk8b/3RVHbPSlyRtnI24T/99wHuTzDO4Zn9Tq98EvKbV3wvs3YBzS5KWsWnlKSurqs8Cn23bTwAXjpjz18AV63E+SdLa+IlcSeqIoS9JHTH0Jakj63JNXwLYtvdTG3r8J6+/bEOPL/XAlb4kdcTQl6SOGPqS1BFDX5I6YuhLUkcMfUnqiKEvSR0x9CWpI4a+JHXE0Jekjhj6ktQRQ1+SOmLoS1JHDH1J6oihL0kd8ffpSzopbOTfY/BvMXyLK31J6oihL0kdWXPoJ9ma5DNJDiZ5LMl7Wv3VSQ4kebw9ntnqSfLhJPNJHk5ywXp9E5KkEzPOSv8I8HNV9f3AxcC1Sc4H9gL3VNV24J62D3ApsL197QFuHOPckqQ1WHPoV9XTVfVg2/5z4CCwGdgF3NKm3QJc3rZ3AbfWwL3AGUnOWXPnkqRVW5dr+km2AW8E7gPOrqqnYfDCALy2TdsMHBp62kKrSZImZOzQT/KdwP8Afqaqvrbc1BG1GnG8PUnmkswtLi6O254kachYoZ/kHzAI/E9U1e+08rNHL9u0x+dafQHYOvT0LcDhpcesqn1VNVtVszMzM+O0J0laYpy7dwLcBBysql8fGtoP7G7bu4E7h+pXt7t4LgZePHoZSJI0GeN8IvdNwE8BjyR5qNV+Ebge+GSSa4CngCva2F3AW4F54CXgnWOcW5K0BmsO/ar6A0ZfpwfYMWJ+Adeu9XySpPH5iVxJ6oihL0kdMfQlqSOGviR1xNCXpI4Y+pLUEUNfkjpi6EtSRwx9SeqIoS9JHTH0Jakjhr4kdcTQl6SOGPqS1BFDX5I6YuhLUkcMfUnqiKEvSR0x9CWpI4a+JHXE0Jekjhj6ktSRiYd+kp1JvpxkPsneSZ9fkno20dBPcgrwEeBS4HzgqiTnT7IHSerZpFf6FwLzVfVEVX0duB3YNeEeJKlbqarJnSx5B7Czqt7V9n8KuKiq3j00Zw+wp+2+HvjyBrZ0FvBnG3j8jWb/02X/0/Vy7n+je/+eqpoZNbBpA086SkbUvu1Vp6r2Afsm0kwyV1WzkzjXRrD/6bL/6Xo59z/N3id9eWcB2Dq0vwU4POEeJKlbkw79+4HtSc5LcipwJbB/wj1IUrcmenmnqo4keTdwN3AKcHNVPTbJHpaYyGWkDWT/02X/0/Vy7n9qvU/0jVxJ0nT5iVxJ6oihL0kdMfQlqSOTvk9/qpJ8H4NPAG9m8PmAw8D+qjo41cY60f77bwbuq6q/GKrvrKrfm15nJybJhUBV1f3t14fsBL5UVXdNubVVS3JrVV097T7WIsmPMfh0/6NV9fvT7mclSS4CDlbV15J8B7AXuAD4IvBfqurFifbTyxu5Sd4HXMXgVz8stPIWBreN3l5V10+rt3EleWdVfWzafSwnyU8D1wIHgTcA76mqO9vYg1V1wTT7W0mS6xj8zqhNwAHgIuCzwE8Cd1fVB6fX3fKSLL0tOsC/AD4NUFVvn3hTq5Dkj6rqwrb97xj8HP0ucAnwP0/2/3eTPAb8cLt7cR/wEnAHsKPV/9VE++ko9P8Y+IGq+saS+qnAY1W1fTqdjS/JU1V17rT7WE6SR4Afqaq/SLKNwQ/9f62qDyX5fFW9caoNrqD1/wbgNOAZYMvQyu2+qvqhqTa4jCQPMlhVfpTBv3AD3MZgwUNV/Z/pdbey4Z+PJPcDb62qxSSvBO6tqn823Q6Xl+RgVX1/2/62BU6Sh6rqDZPsp6fLO98Evhv4kyX1c9rYSS3Jw8cbAs6eZC9rdMrRSzpV9WSSNwN3JPkeRv96jpPNkar6W+ClJP+/qr4GUFV/leRk//mZBd4DfAD4+ap6KMlfnexhP+QVSc5k8B5kqmoRoKr+MsmR6bZ2Qh4d+tf4F5LMVtVckn8KfGOlJ6+3nkL/Z4B7kjwOHGq1c4HXAe8+7rNOHmcDbwFeWFIP8P8m386qPZPkDVX1EEBb8b8NuBk4qVdqzdeTnF5VLwH//Ggxyas4yRcNVfVN4IYk/709PsvL6//9VwEPMPhZryTfVVXPJPlOXh4LhncBH0rynxn8krU/THKIQQ69a9LNdHN5ByDJKxi8AbSZwQ/LAnB/W8Gd1JLcBHysqv5gxNh/q6p/PYW2TliSLQxWy8+MGHtTVf3fKbR1wpKcVlV/M6J+FnBOVT0yhbbWJMllwJuq6hen3cs4kpwOnF1VX5l2LyciyT8CvpfBC+5CVT07lT56Cn1J6p336UtSRwx9SeqIoS9JHTH0Jakjhr4kdeTvAL8g9bx+H0KrAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "data['class'].value_counts().sort_index().plot.bar()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Spliiting the data into testing and training set to perform analysis of given DNA sequences "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.model_selection import train_test_split \n",
    "x_train, x_test, y_train, y_test= train_test_split(x,y_data, test_size=.20,random_state=40)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(876, 232414)\n",
      "(3504, 232414)\n"
     ]
    }
   ],
   "source": [
    "print(x_test.shape)\n",
    "print(x_train.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "using Multinomial Bayes claassifier with grid value of 1 (obtained from various experiments)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MultinomialNB(alpha=0.1, class_prior=None, fit_prior=True)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#multinomial bayes classifier\n",
    "from sklearn.naive_bayes import MultinomialNB\n",
    "MNB  = MultinomialNB(alpha = 0.1)\n",
    "MNB.fit(x_train, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "y_pred = MNB.predict(x_test)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "User defined function to calculate the accuracy of the model deployed. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Confusion Matrix\n",
      "\n",
      "predicted    0    1   2    3    4   5    6\n",
      "Actual                                    \n",
      "0          113    0   0    0    0   0    0\n",
      "1            1  115   0    0    0   0    4\n",
      "2            0    0  62    0    0   0    3\n",
      "3            0    0   0  121    1   0    1\n",
      "4            1    0   0    0  149   0    2\n",
      "5            2    0   0    0    0  37    1\n",
      "6            0    0   0    0    0   0  263\n",
      "\n",
      "Accuracy Score = 0.982\n",
      "Precision Score = 0.982\n",
      "Recall Score = 0.982\n",
      "F1 Score = 0.982\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import accuracy_score, f1_score, precision_score, recall_score\n",
    "print('Confusion Matrix\\n')\n",
    "print(pd.crosstab(pd.Series(y_test,name='Actual'),pd.Series(y_pred, name ='predicted')))\n",
    "def get_metric(y_test,y_pred):\n",
    "    accuracy=accuracy_score(y_test,y_pred)\n",
    "    precision = precision_score(y_test, y_pred, average='weighted')\n",
    "    recall = recall_score(y_test, y_pred, average='weighted')\n",
    "    f1 = f1_score(y_test, y_pred, average='weighted')\n",
    "    return accuracy, precision, recall, f1\n",
    "accuracy,precision, recall, f1 = get_metric(y_test, y_pred)\n",
    "print('\\nAccuracy Score = %.3f\\nPrecision Score = %.3f\\nRecall Score = %.3f\\nF1 Score = %.3f'% (accuracy,precision, recall, f1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Inference:\n",
    "1. from the given dataset, It is evident that the classes which the DNA Sequences are labled are essential factors while forming the k mers. \n",
    "2. The model didn't overfit/underfit with the dataset, as the score of accuracy is moderatly high for unseen data. \n",
    "3. Depending on the gene/codon sequence to be mapped, the lenth of K mer can be changed to any arbitrary whole number resulting in different sequences/combinations. \n",
    "4. The above methid is usefull in finding the critical points in DNA sequences such as binding and transcription sites.  "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
